torch/_inductor/lowering.py

```diff
@@ -4276,60 +4269,53 @@ def _avg_poolnd(
-            x,
-            kernel_size,
-            stride,
-            padding,
-            ceil_mode,
-            count_include_pad,
-            divisor_override,
+    def fn_sum(loader):
+        def fn_inner(idx, reduction_idx):
+            prefix = idx[:-dim]
+            bh = idx[-dim:]
+            ih = reduction_idx
+            ih = [bh[i] * stride[i] + ih[i] - padding[i] for i in range(dim)]
+            return loader([*prefix, *ih])
+
+        rv = Reduction.create(
+            reduction_type="sum",
+            input_node=x,
+            device=x.get_device(),
+            dst_dtype=dtype,
+            src_dtype=dtype,
+            inner_fn=fn_inner,
+            ranges=new_size,
+            reduction_ranges=kernel_size,
         )
-
-    def fn_sum(idx, loader):
-        prefix = idx[:-dim]
-        b = idx[-dim:]
-        total = None
-        for ih in itertools.product(*[range(kernel_size[i]) for i in range(dim)]):
-            inp = [b[i] * stride[i] + ih[i] - padding[i] for i in range(dim)]
-            val = loader([*prefix, *inp])
-            if total is None:
-                total = val
-            else:
-                total = ops.add(val, total)
-        return total
+        if isinstance(rv.data.data, Reduction):
+            # Only realize if reduction isn't unrolled
+            rv.realize()
+        return rv
 
     if not had_padding or divisor_override:
         if divisor_override:
             scale = 1 / divisor_override
         else:
-            scale = 1.0 / window_size
-
-        def fn(idx):
-            return ops.mul(fn_sum(idx, x_loader), ops.constant(scale, dtype))
+            scale = 1.0 / functools.reduce(operator.mul, kernel_size)
 
+        result = mul(fn_sum(x_loader), scale)
     else:
-        ones_loader = constant_boundary_condition(
-            ones_like(x),
-            1.0 if count_include_pad else 0.0,
-            dim=dim,
-        )
 
-        def fn(idx):
-            # TODO(jansel): optimize to do `int(x<h)` rather than `x<h?1:0`
-            return ops.truediv(fn_sum(idx, x_loader), fn_sum(idx, ones_loader))
+        def ones_loader(idx):
+            prefix = idx[:-dim]
+            bh = idx[-dim:]
+            lower = [-padding[i] if count_include_pad else 0 for i in range(dim)]
+            upper = [
+                h[i] + padding[i] if count_include_pad else h[i] for i in range(dim)
+            ]
+            masks = [range_mask(bh[i], upper[i], lower[i]) for i in range(dim)]
+            mask = functools.reduce(ops.and_, masks)
+            return ops.masked(mask, lambda: ops.constant(1, dtype), 0)
 
-    rv = Pointwise.create(
-        device=x.get_device(),
-        dtype=dtype,
-        inner_fn=fn,
-        ranges=new_size,
-    )
-    # TODO(jansel): should we force these to be realized?
-    return rv
+        # TODO(jansel): optimize to do `int(x<h)` rather than `x<h?1:0`
+        result = div(fn_sum(x_loader), fn_sum(ones_loader))
```

Comment:
For large kernel sizes I expect `sum(ones_loader)` is quite inefficient. Can we write it as an index_expr instead? Something like using min and max to get the in-bounds shape in each dim and multiply them together to get the volume.

Reply:
I did write such a function and the runtime came out to be slower than this version.

Reply:
Do you still have the code? I'd be curious to see why it was worse.

Reply:
I was fiddling around with reduction hints at the same time and `sum(ones_loader)` was better on one of those hints, but that hint was not good anyway. I've pushed a commit.